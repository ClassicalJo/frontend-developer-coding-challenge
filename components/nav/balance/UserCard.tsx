import React, { SyntheticEvent, useState } from "react";
import fetchPoints from "../../fetchPoints";
import keyDown from "../../keyDown";
import { UserData, ValidCharge } from "../../types";
import Aerocard from "./Aerocard";
import {
    StyledUserCardContainer, StyledUserCardPointsSelect, StyledUserCardPointsSelectWrapper, StyledUserCardPointsWrapper,
    StyledUserCardSection, StyledUserCardTitle
} from "./styles";
import UserCardPointsButton from "./UserCardPointsButton";

interface AppProps {
    hide: Boolean;
    userData: UserData;
    refreshUserData: () => void;
    successToast: (item: string) => void;
    errorToast: () => void;
}
export default function UserCard({ hide, userData, refreshUserData, successToast, errorToast }: AppProps): JSX.Element {
    let [currentCharge, setCurrentCharge] = useState<ValidCharge>(1000)
    let [loading, setLoading] = useState<Boolean>(false)
    let selectValues: ValidCharge[] = [1000, 5000, 7500]
    function generateHandleClick(k: ValidCharge) {
        return function (e: React.MouseEvent) {
            e.stopPropagation()
            setCurrentCharge(k)
        }
    }
    function generateHandleKeyDown(k: ValidCharge) {
        return function (e: React.KeyboardEvent) {
            e.stopPropagation()
            setCurrentCharge(k)
        }
    }
    function successfulCharge(charge: number) {
        refreshUserData()
        successToast(charge + " Aeropoints package")
    }
    function chargePoints() {
        let charge = currentCharge
        setLoading(true)
        fetchPoints(charge)
            .then(() => successfulCharge(charge))
            .catch(() => errorToast())
            .finally(() => setLoading(false))
    }
    return (
        <StyledUserCardContainer hide={hide} onClick={(e: SyntheticEvent) => e.stopPropagation()}>
            <StyledUserCardTitle>Add Balance</StyledUserCardTitle>
            <StyledUserCardSection>
                <Aerocard userData={userData} />
                <StyledUserCardPointsWrapper >
                    {selectValues.map((k, i) => (
                        <StyledUserCardPointsSelectWrapper
                            key={`userCardSelectPoints${i}`}
                            tabIndex={0}
                            current={currentCharge === k}
                            role='button'
                            onClick={generateHandleClick(k)}
                            onKeyDown={keyDown(generateHandleKeyDown(k))}>
                            <StyledUserCardPointsSelect current={currentCharge === k}>
                                <p>{k}</p>
                            </StyledUserCardPointsSelect>
                        </StyledUserCardPointsSelectWrapper>
                    ))}
                </StyledUserCardPointsWrapper>
                <UserCardPointsButton loading={loading} chargePoints={chargePoints} />
            </StyledUserCardSection>
        </StyledUserCardContainer >
    )
}