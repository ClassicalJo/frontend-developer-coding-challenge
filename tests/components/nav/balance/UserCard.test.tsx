import UserCard from "@nav/balance/UserCard";
import { act, render, waitFor } from "@testing-library/react";
import { ValidCharge } from "@types";
import { FAKE_USER_DATA } from "../../../setupServer";
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'

describe("Testing the UserCard component", () => {
    let props = {
        hide: false,
        userData: FAKE_USER_DATA,
        refreshUserData: jest.fn(),
        successToast: jest.fn(),
        errorToast: jest.fn(),
    }
    it("Should render: Add balance", () => {
        let component = render(<UserCard {...props} />)
        expect(component.container).toHaveTextContent("Add Balance")
    })
    it("Should render Aerocard component", () => {
        let component = render(<UserCard {...props} />)
        expect(component.container).toHaveTextContent("Aerocard")
    })
    it("Should render three charge option buttons", () => {
        let component = render(<UserCard {...props} />)
        let validCharges: ValidCharge[] = [1000, 5000, 7500]
        validCharges.forEach(k => {
            let button = component.getByRole('button', { name: k.toString() })
            expect(button).toBeInTheDocument()
        })
    })
    it("Should render a redeem button", () => {
        let component = render(<UserCard {...props} />)
        let button = component.getByRole('button', { name: "Add Points" })
        expect(button).toBeInTheDocument()
    })
    it("Clicking the redeem button should change it to processing", async () => {
        let component = render(<UserCard {...props} />)
        let button = component.getByRole('button', { name: "Add Points" })
        let user = userEvent.setup()
        await user.click(button)
        expect(component.container).toHaveTextContent("Processing...")
    })
    it("Clicking the add points button should make an api call and generate a toast", async () => {
        let component = render(<UserCard {...props} />)
        let button = component.getByRole('button', { name: "Add Points" })
        let user = userEvent.setup()
        await user.click(button)
        expect(props.successToast).toBeCalledWith("1000 Aeropoints package")
    })
    it("Clicking a charge button then the Add Points button should make a toast for the points", async () => {
        let component = render(<UserCard {...props} />)
        let amount: ValidCharge = 7500
        let user = userEvent.setup()

        let chargeButton = component.getByRole('button', { name: amount.toString() })
        await user.click(chargeButton)
        let redeemButton = component.getByRole('button', { name: "Add Points" })
        await user.click(redeemButton)
        await waitFor(() => {
            expect(props.successToast).toBeCalledWith(amount + " Aeropoints package")
        })
    })
    it("Clicking a charge button different from the currently selected should change its style", async () => {
        let component = render(<UserCard {...props} />)
        let amount: ValidCharge = 5000
        let user = userEvent.setup()
        let chargeButton = component.getByRole('button', { name: amount.toString() })
        await user.click(chargeButton)
        expect(chargeButton).toMatchSnapshot()
    })

})