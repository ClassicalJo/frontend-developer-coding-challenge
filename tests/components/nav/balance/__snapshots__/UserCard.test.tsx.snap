// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the UserCard component Clicking a charge button different from the currently selected should change its style 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 81px;
  height: 35px;
  border-radius: 12px;
  background: linear-gradient(102.47deg,#176FEB -5.34%,#FF80FF 106.58%);
  -webkit-animation: colorBackground .15s forwards;
  animation: colorBackground .15s forwards;
}

.c1 {
  color: white;
}

@media (min-width:1920px) {

}

@media (min-width:1920px) {

}

@media (min-width:1920px) {

}

<div
  class="c0"
  role="button"
  tabindex="0"
>
  <div
    class="c1"
  >
    <p>
      5000
    </p>
  </div>
</div>
`;
